code/mission/import/xwingmissionparse.cpp

```diff
@@ -0,0 +1,743 @@
+#include "iff_defs/iff_defs.h"
+#include "mission/missionparse.h"
+#include "mission/missiongoals.h"
+#include "missionui/redalert.h"
+#include "nebula/neb.h"
+#include "parse/parselo.h"
+#include "ship/ship.h"
+#include "species_defs/species_defs.h"
+#include "starfield/starfield.h"
+
+#include "xwingbrflib.h"
+#include "xwinglib.h"
+#include "xwingmissionparse.h"
+
+extern int allocate_subsys_status();
+
+static int Player_flight_group = 0;
+
+// vazor222
+void parse_xwi_mission_info(mission *pm, const XWingMission *xwim)
+{
+	pm->author = "X-Wing";
+	strcpy_s(pm->created, "00/00/00 at 00:00:00");
+
+	// NOTE: Y and Z are swapped and the units are in km
+	Parse_viewer_pos.xyz.x = 1000 * xwim->flightgroups[Player_flight_group].start1_x;
+	Parse_viewer_pos.xyz.y = 1000 * xwim->flightgroups[Player_flight_group].start1_z + 100;
+	Parse_viewer_pos.xyz.z = 1000 * xwim->flightgroups[Player_flight_group].start1_y - 100;
+	vm_angle_2_matrix(&Parse_viewer_orient, PI_4, 0);
+}
+
+bool is_fighter_or_bomber(const XWMFlightGroup *fg)
+{
+	switch (fg->flightGroupType)
+	{
+		case XWMFlightGroupType::fg_X_Wing:
+		case XWMFlightGroupType::fg_Y_Wing:
+		case XWMFlightGroupType::fg_A_Wing:
+		case XWMFlightGroupType::fg_B_Wing:
+		case XWMFlightGroupType::fg_TIE_Fighter:
+		case XWMFlightGroupType::fg_TIE_Interceptor:
+		case XWMFlightGroupType::fg_TIE_Bomber:
+		case XWMFlightGroupType::fg_Gunboat:
+		case XWMFlightGroupType::fg_TIE_Advanced:
+			return true;
+		default:
+			break;
+	}
+	return false;
+}
+
+bool is_wing(const XWMFlightGroup *fg)
+{
+	return (fg->numberInWave > 1 || fg->numberOfWaves > 1 || is_fighter_or_bomber(fg));
+}
+
+int xwi_flightgroup_lookup(const XWingMission *xwim, const XWMFlightGroup *fg)
+{
+	for (size_t i = 0; i < xwim->flightgroups.size(); i++)
+	{
+		if (xwim->flightgroups[i].designation == fg->designation)
+			return (int)i;
+	}
+	return -1;
+}
+
+void xwi_add_attack_check(const XWingMission *xwim, const XWMFlightGroup *fg)
+{
+	char fg_name[NAME_LENGTH] = "";
+	char event_name[NAME_LENGTH];
+	char sexp_buf[NAME_LENGTH + 50];
+
+	int fg_index = xwi_flightgroup_lookup(xwim, fg);
+	Assertion(fg_index >= 0, "Flight Group index must be valid");
+
+	strcpy_s(fg_name, fg->designation.c_str());
+	SCP_totitle(fg_name);
+
+	sprintf(event_name, "FG %d Attack Check", fg_index);
+
+	if (mission_event_lookup(event_name) >= 0)
+		return;
+
+	Mission_events.emplace_back();
+	auto event = &Mission_events.back();
+	event->name = event_name;
+
+	if (is_wing(fg))
+		sprintf(sexp_buf, "( when ( true ) ( fotg-wing-attacked-init \"%s\" ) )", fg_name);
```

Comment:
hang on, is this an actual sexp?

I know why this is here, but I am strictly against encoding a dependency on a modpack this way. Why isn't this part of some translation table that could be shipped separately?

Reply:
It is a scripted sexp in the FotG modpack.

I know what you mean, having a dependency on a mod is a code smell, but this is one of the rare times it is justified.  The XWI importer needs to process a lot of data in several very specific ways.  It isn't generalizable.  You can't create a table for that sort of information, not without it requiring a lot of pointlessly over-engineered options like this:
```
$FG Attack Check Event: FG %d Attack Check
+IfWing: ( when ( true ) ( fotg-wing-attacked-init "%s" ) )
+IfShip: ( when ( true ) ( fotg-ship-attacked-init \"%s\" ) )
```
If you know, from the task requirements, that you will only ever have one use case, then it makes sense to handle that use case as directly and simply as possible.  Creating a bunch of arbitrary hoops to jump through is just doing work for no benefit.  Premature complexity should be avoided even more strongly than premature optimization.

The only parts of the importer that could plausibly be specified in a table are the lists: the ship classes, ranks, formations, and whether a ship is a fighter or bomber.  Everything else involves a great deal of moving parts that have to fit together in a specific way and usually in a specific order.  That sort of code belongs in the importer.  It can't be done in a table, and it can't even be done in a script.